Document how to link loadable shared libraries with Celeritas

drbenmorgan opened the pull request:
Integration of Celeritas in ATLAS's FullSimLight required it to be used in and linked from a shared library loaded by the FullSimLight application at runtime (a "plugin"). `celeritas_target_link_libraries` on the plugin library did not result in the needed "_final" device code library being linked to it, leading to undefined symbols at load time. Traced to `celeritas_target_link_libraries` not considering a shared library as an endpoint required full device link resolution.

As requested in offline discussion, this adds some brief documentation on this use case, and the workaround. I'm not sure if it's too verbose, or equally not enough detail 😆, but easy to address this in review.

drbenmorgan committed:
Document how to link loadable shared libraries with Celeritas

Integration of Celeritas in ATLAS's FullSimLight required it to
be used in and linked from a shared library loaded by the FullSimLight
application at runtime (a "plugin"). `celeritas_target_link_libraries`
on the plugin library did not result in the needed "_final" device code
library being linked to it, leading to undefined symbols at load time.
Traced to `celeritas_target_link_libraries` not considering a shared
library as an endpoint required full device link resolution.

Briefly describe this use case in documentation and provide example
of how to link a "plugin" shared library to Celeritas targets.

sethrj reviewed (approved):
The documentation looks perfect, but perhaps it would be better still if you and/or @pcanal could add an extension to `celeritas_target_link_libraries` to query the target's `TYPE` property and if it's `MODULE_LIBRARY` add the correct targets?

pcanal commented:
> could add an extension to celeritas_target_link_libraries to query the target's TYPE property and if it's MODULE_LIBRARY add the correct targets?

Indeed this could be done :)

sethrj commented:
@pcanal If you can fix the module linking in a new pull request, then Ben could merge develop in to this branch, make sure it works, and update this documentation to note that the `link_libraries` macro works when the target is a module as well?

pcanal commented:
The fix for this exact case turned out to be easy.   One (new) left over issue is what to do if the MODULE library contains RDC cuda.   Typically that triggers the split in the 3 libraries (static and shared for down-stream and final library).   Since the MODULE library should have no downstream it feel likes we should only produce the final library (with the original name)  ... the "only" issue is then the somewhat 'irregular' set up for those libraries.

drbenmorgan commented:
Above all sounds good - once @pcanal's fix is in I'll rebase here to update the text to describe `MODULE` support.

pcanal commented:
The related PR https://github.com/celeritas-project/celeritas/pull/848 has been merged.

sethrj commented:
@drbenmorgan At your convenience can you merge develop into your branch and update the instructions accordingly? It sounds like there should be very little that needs to be done now. Thanks!

pcanal commented:
@sethrj Actually given the fix the instruction are no longer the "best" and can be simplified.

pcanal commented on doc/usage.rst:
```diff
@@ -39,6 +39,30 @@ of ``target_link_libraries`` with a customized version::
+Celeritas, or Celeritas consumers, in your project to track the appropriate sequence
+of linking for the final application whether CPU-only or CUDA enabled::
+
+  add_library(myconsumer SHARED ...)
```
Note that the same description given above for ``celeritas_target_link_libraries`` is also true for ``celeritas_add_library``
```suggestion
  celeritas_add_library(myconsumer SHARED ...)
```
Furthermore if the library contains CUDA code, using celeritas_add_library is compulsory (I forgot if you can mix RDC library with non RDC library which might be the only case were you might not need it).

drbenmorgan replied:
Good call. on this, I've updated this in all of the locations touched here.

pcanal commented on doc/usage.rst:
```diff
@@ -39,6 +39,30 @@ of ``target_link_libraries`` with a customized version::
+  add_library(myconsumer SHARED ...)
+  celeritas_target_link_libraries(myconsumer PUBLIC Celeritas::celeritas)
+
+  add_executable(myapplication ...)
```
Same applies to ``celeritas_add_executable``

pcanal commented on doc/usage.rst:
```diff
@@ -39,6 +39,30 @@ of ``target_link_libraries`` with a customized version::
+  add_executable(myapplication ...)
+  celeritas_target_link_libraries(myapplication PRIVATE myconsumer)
+
+The one exception to this is when your project builds a shared library that will
```
I think with the fix this is no longer needed.

drbenmorgan replied:
I've left this in but updated to recommend `MODULE` over `SHARED`, leaving the workaround for cases where `MODULE` cannot be used (think there are quite a few projects that don't/won't use `MODULE`...).

pcanal replied:
If you use a "SHARED" library as a plugin I would still use the `celeritas_` library as it would still properly handle if that library contains cuda code (i.e. the text 'works' only if the library does not contain any cuda code).

The only to note is that they should then load the "_final" library as their plugin.

drbenmorgan replied:
The case I was thinking about here is if a project is providing a custom `add_library` of their own, e.g. `myprojects_add_library` and similar. The secondary issue is that certain systems may enforce a naming convention for plugins, and so loading the `_final` library may not be possible without a softlink or similar. I can clarify both of these.

.. edit.. added this in latest commit.

pcanal replied:
Technically they should change `myprojects_add_library` to use `celeritas_add_library` :) but fair enough it makes sense and the hints is useful anyway (if only to talk about the `_final` library.

drbenmorgan committed:
Clarify shared/module library linking after RDC update

PR #848

drbenmorgan committed:
Clarify shared/module library linking after RDC update

PR #848 taught celeritas_target_link_libraries that MODULE targets
are always the final one in any link chain and hence to device link
them correctly.

Update text on MODULE vs SHARED targets to account for this update.
Retain text on explicit linking of final target to cover case of
projects that enforce use of SHARED targets for plugins.

drbenmorgan committed:
Address review comments on Celeritas CMake wrappers

Show use of Celeritas wrappers in all places, noting decay properties
allowing any combination of CUDA/CPU compiling and linking.

drbenmorgan commented:
There is one corner case not documented here, though am pretty sure it won't bite in the near term, which is:

1. Project "A" uses Celeritas in a `SHARED` library it installs.
2. Project "B" links to "A"s shared library

"A" either needs to fully device link that library, or document that consumers need to use `celeritas_target_link_libraries`. Can be documented, but don't think it's critical for now, and probably better addressed by upstreaming RDC support to CMake.

pcanal commented:
Note, I intent to upload (hopefully this week) the splitting between the `celertias_` function and the 'really' generic `rdc_` functions.

pcanal commented:
> probably better addressed by upstreaming RDC support to CMake.

True.  By the way I get up to date my fork: https://github.com/pcanal/CudaRDCExercise

> Can be documented, but don't think it's critical for now,

We should recommend they point to that/this document.

sethrj reviewed (approved):
Looks great! Thanks @drbenmorgan for the documentation and figuring out how to get plugins working.

drbenmorgan committed:
Clarify use cases for linking to SHARED plugin library

Celeritas recommends use of celeritas_add_library to support the full
RDC chain but it may not be possible to use this in projects that
provide their own wrappers over the raw CMake commands.

Add clarification on the use case here, retaining example use of raw
CMake commands for simplest possible use case.

pcanal commented:
> We should recommend they point to that/this document.

Did we add this somewhere?
